hash.go

```diff
@@ -358,6 +346,23 @@ func (h *md5Marshal) UnmarshalBinary(b []byte) error {
 	return nil
 }
 
+func (h *md5Marshal) AppendBinary(buf []byte) ([]byte, error) {
+	d := (*md5State)(h.hashState())
+	if d == nil {
+		return nil, errors.New("crypto/md5: can't retrieve hash state")
+	}
+
+	buf = append(buf, md5Magic...)
+	buf = appendUint32(buf, d.h[0])
+	buf = appendUint32(buf, d.h[1])
+	buf = appendUint32(buf, d.h[2])
+	buf = appendUint32(buf, d.h[3])
+	buf = append(buf, d.x[:d.nx]...)
+	buf = buf[:len(buf)+len(d.x)-int(d.nx)] // already zero
```

Comment:
This seems more robust: `b = append(b, make([]byte, len(d.x)-d.nx)...)`. Same for other marshalers.

Reply:
fixed